src/utils/intrusive_ptr.h

```diff
@@ -0,0 +1,81 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#ifndef SRC_UTILS_INTRUSIVE_PTR_H_
+#define SRC_UTILS_INTRUSIVE_PTR_H_
+
+#include "src/utils/ref_counted.h"
+
+namespace raksha {
+
+template <typename T>
+class intrusive_ptr {
+ public:
+  intrusive_ptr() : ptr_(nullptr) {}
+  intrusive_ptr(T* ptr) : ptr_(Retain(ptr)) {}
+  intrusive_ptr(const intrusive_ptr<T>& other) : ptr_(Retain(other.ptr_)) {}
+  // There is no need to adjust count on a move.
+  intrusive_ptr(intrusive_ptr<T>&& other) : ptr_(other.ptr_) {
+    other.ptr_ = nullptr;
+  }
+  ~intrusive_ptr() { Release(ptr_); }
+
+  intrusive_ptr<T>& operator=(intrusive_ptr<T> other) {
+    // We are using the copy-and-swap idiom for assignment.
+    // Argument `other` should be passed by value.
+    this->Swap(other);
+    return *this;
+  }
+
+  T& operator*() const {
+    CHECK(ptr_ != nullptr) << "Dereferencing a nullptr!";
+    return *ptr_;
+  }
+
+  T* operator->() const { return ptr_; }
+  T* get() const { return ptr_; }
```

Comment:
I wonder whether we need `get()`. It seems like it could undermine our proof that no uncounted copies of the reference exist. It looks like this is mostly used for comparing equality in tests; can we make a `PointersEqual` method instead so as not to exflitrate the contained pointer?

Reply:
Good point.  Let me address this in a follow up PR as I want to add pointer comparators anyway. 

In any case, even if we remove the `get` method, we can still get the raw pointer value using `&*p`.